src/main/java/org/jabref/logic/importer/FulltextFetchers.java

```diff
@@ -36,26 +48,93 @@ public FulltextFetchers(ImportFormatPreferences importFormatPreferences) {
         Optional<DOI> doi = clonedEntry.getField(FieldName.DOI).flatMap(DOI::parse);
 
         if (!doi.isPresent()) {
+            findDoiForEntry(clonedEntry);
+        }
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try {
+            List<Future<Optional<FetcherResult>>> result = new ArrayList<>();
             try {
-                WebFetchers.getIdFetcherForIdentifier(DOI.class)
-                        .findIdentifier(clonedEntry)
-                        .ifPresent(e -> clonedEntry.setField(FieldName.DOI, e.getDOI()));
-            } catch (FetcherException e) {
-                LOGGER.debug("Failed to find DOI", e);
+                result = executor.invokeAll(getCallables(clonedEntry, finders), FETCHER_TIMEOUT, TimeUnit.SECONDS);
+            } catch (InterruptedException ignore) {
+
             }
+
+            return result.stream()
+                    .map(FulltextFetchers::waitForResult)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .filter(res -> Objects.nonNull(res.source))
```

Comment:
In my opinion, the fetcher should make sure that the URL returned is always non-null. Otherwise the usage of `Optional` is superfluous.